rfc/001-composable-pipelines.md

```diff
@@ -0,0 +1,392 @@
+| `from`   | `Void`     | `Bytes`     | Loads bytes from a source as a side effect
+| `to`     | `Bytes`    | `Void`      | Writes bytes to a sink as a side effect
+| `read`   | `Bytes`    | `Arrow`     | Parses a specific format into Arrow
+| `write`  | `Arrow`    | `Bytes`     | Prints input in a specific format
+| `pull`   | `Void`     | `Arrow`     | Loads Arrow from a remote source
+| `push`   | `Arrow`    | `Void`      | Stores Arrow in a remote source
+
+#### Compute
+
+The following operators have a fixed input and output type of `Arrow`. We can
+group them into *filter* operators that do not change the schema, and *reshape*
+operators where execution yields a new schema.
+
+##### Filter
+
+Filter operators have input and output type `Arrow`.
+
+| Operator | Non-Blocking | Description
+| -------- |:------------:| ------------------------------------------
+| `head`   | ✅           | Retrieve first N records
+| `tail`   | ❌           | Retrieve last N records
+| `where`  | ✅           | Filter with expression
+| `sort`   | ❌           | Sort according to field
+| `uniq`   | ✅           | Produce deduplicated output
+
+##### Reshape
+
+Reshape operators have input and output type `Arrow`.
+
+| Operator    | Non-Blocking | Description
+| ----------- |:------------:| ------------------------------------------
+| `put`       | ✅           | Select a set of columns (projection)
+| `drop`      | ✅           | Remove a set of columns (projection)
+| `replace`   | ✅           | Replaces field described by extractors with values
+| `extend`    | ✅           | Adds new fields with (initially fixed) values
+| `rename`    | ✅           | Renames schema meta data (type & fields)
+| `summarize` | ❌/✅[^1]    | Aggregate group and compute a summary function
+
+[^1]: Depending on the aggregate function.
+
+##### Matcher Plugin
+
+| Operator | Input Type | Output Type | Description
+| -------- | ---------- | ----------- | ------------------------------------------
+| `build`  | `Arrow`    | `Matcher`   | Creates a matcher from a data stream
+| `match`  | `Arrow`    | `Arrow`     | Matches input based on its state
+
+#### Commands
+
+The following table illustrate how existing VAST commands map to pipeline
+execution.
+
+| Command Synatx                    | Pipeline Syntax
+| --------------------------------- | -------------------------------------------
+| `vast infer`                      | `vast exec 'read | infer`
+| `vast import FORMAT EXPR`         | `vast exec 'read FORMAT | where EXPR | push'`
+| `vast export FORMAT EXPR`         | `vast spawn 'where EXPR | write FORMAT'`
+| `vast count [-e] EXPR`            | `vast spawn 'where EXPR | count [-e]` (TBD)
+| `vast pivot --format=FORMAT EXPR` | `vast spawn 'where EXPR | pivot | write FORMAT`
+| `vast explore [-A] [...] EXPR`    | `vast spawn 'where EXPR | explore [-A] [...]`
+
+The `spawn` command create a remote pipeline and attaches immediately to it.
```

Comment:
Not sure I understand how spawn relates to pull and push

Reply:
The idea is that there is a dual of `exec` and `spawn`, one being local execution and the other remote. The latter means (1) create the pipeline, and (2) attach to it so that you still get the data from remote.

Example:

```bash
vast spawn PIPELINE
vast exec PIPELINE
```

With `spawn`, we instantiate the pipeline remotely. With `exec`, we run it in situ and block the process until receiving SIGINT.

I see `push` and `pull` as a way to "hydrate" and "rehydrate" data at a specified VAST node.

Reply:
I am still a bit confused, sorry 😄. Could you explicit the difference between:
- `vast spawn 'where EXPR | write FORMAT'`
- `vast pull 'where EXPR | write FORMAT'`

Reply:
Given the confusion, this will go into a separate proposal. Will resolve this for now.